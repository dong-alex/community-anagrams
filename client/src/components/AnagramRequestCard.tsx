import React, {
  useState,
  useEffect,
  FunctionComponent,
  FormEvent,
} from "react";
import {
  MDBBtn,
  MDBCard,
  MDBCardBody,
  MDBCardText,
  MDBInput,
  MDBTypography,
} from "mdbreact";
import { AnagramSubmission } from "../types";
import isAlphabetical from "./utils/isAlphabetical";
import useAnagrams from "./hooks/useAnagrams";

const AnagramRequestCard: FunctionComponent<{}> = () => {
  const { onSaveAnagram } = useAnagrams();
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [resultMessage, setResultMessage] = useState<string>("");
  const [disableSubmission, setDisableSubmission] = useState<boolean>(false);
  const [submission, setSubmission] = useState<AnagramSubmission>({
    firstWord: "",
    secondWord: "",
  });

  useEffect(() => {
    const { firstWord, secondWord } = submission;
    if (
      (firstWord !== "" && !isAlphabetical(firstWord)) ||
      (secondWord !== "" && !isAlphabetical(secondWord))
    ) {
      setErrorMessage("Please remove any numbers or symbols.");
      setDisableSubmission(true);
    } else {
      setErrorMessage("");
    }

    setDisableSubmission(firstWord === "" || secondWord === "");
  }, [submission]);

  const handleChange = (event: FormEvent<HTMLInputElement>): void => {
    setSubmission({
      ...submission,
      [event.currentTarget.name]: event.currentTarget.value,
    });
  };

  const handleSubmitAnagramCheck = async (
    event: FormEvent<HTMLButtonElement>
  ): Promise<void> => {
    event.preventDefault();
    const { firstWord, secondWord } = submission;

    // useEffect handled validations
    try {
      const result = await onSaveAnagram(firstWord, secondWord);
      handleDisplayResult(result);
      return;
    } catch (err) {
      console.log("Error with handleSubmitAnagramCheck", err);
      setErrorMessage(
        "There was an issue with checking the words, please try again."
      );
    }
  };

  const handleDisplayResult = (result: boolean): void => {
    // copy the initial words so that we can check anagrams again
    const { firstWord, secondWord } = submission;

    let savedFirstWord = firstWord;
    let savedSecondWord = secondWord;

    const successString = `'${savedFirstWord}' and '${savedSecondWord}' are not anagrams. Feel free to check other words and see if they are anagrams.`;
    const failureString = `Unfortunately '${savedFirstWord}' and '${savedSecondWord}' are not anagrams. Feel free to check other words and see if they are anagrams.`;

    setResultMessage(result ? successString : failureString);
  };

  return (
    <MDBCard className="mt-5">
      <MDBCardBody>
        <MDBCardText>
          <form>
            <MDBTypography
              tag="h4"
              variant="h4-responsive"
              className="text-center mb-4"
            >
              Check Anagram
            </MDBTypography>
            {resultMessage !== "" && (
              <div
                className="alert alert-dark text-center"
                data-test-id="result-alert"
                role="alert"
              >
                {resultMessage}
              </div>
            )}
            <p className="text-center ">
              Check if two words are anagrams here.{" "}
              <strong>Do not include special characters and numbers.</strong>
            </p>
            {errorMessage !== "" && (
              <div
                className="alert alert-danger text-center"
                data-test-id="error-alert"
                role="alert"
              >
                {errorMessage}
              </div>
            )}
            <div className="grey-text">
              <MDBInput
                id="first-word-input"
                data-test-id="first-word-input"
                name="firstWord"
                onChange={handleChange}
                label="First Word"
                group
                type="text"
              />
              <MDBInput
                id="second-word-input"
                data-test-id="second-word-input"
                name="secondWord"
                onChange={handleChange}
                label="Second Word"
                group
                type="text"
              />
            </div>
            <div className="text-center">
              <MDBBtn
                onClick={handleSubmitAnagramCheck}
                data-test-id="anagram-submission-button"
                color="blue"
                disabled={disableSubmission}
              >
                Check
              </MDBBtn>
            </div>
          </form>
        </MDBCardText>
      </MDBCardBody>
    </MDBCard>
  );
};

export default AnagramRequestCard;
