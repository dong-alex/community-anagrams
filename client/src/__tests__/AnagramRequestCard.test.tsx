import React from "react";
import { render, screen, act, fireEvent } from "@testing-library/react";
import AnagramRequestCard from "../components/AnagramRequestCard";

describe("AnagramRequestCard Tests", () => {
  it("renders the basic card", () => {
    const stub = jest.fn();

    const { container } = render(<AnagramRequestCard onSaveAnagram={stub} />);

    expect(container).toMatchSnapshot();
  });

  it("calls the api with the correct parameters - Not an Anagram", async () => {
    const stub = jest.fn((fw: string, sw: string) => Promise.resolve(false));

    render(<AnagramRequestCard onSaveAnagram={stub} />);

    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "firstword", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "secondword", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(stub).toHaveBeenCalledWith("firstword", "secondword");

    expect(screen.getByTestId("result-alert")).toMatchSnapshot(); // snapshot TODO
  });

  it("calls the api with the correct parameters - Is An Anagram", async () => {
    const stub = jest.fn((fw: string, sw: string) => Promise.resolve(true));

    render(<AnagramRequestCard onSaveAnagram={stub} />);

    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "racer", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "carer", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(stub).toHaveBeenCalledWith("racer", "carer");
    expect(screen.getByTestId("result-alert")).toMatchSnapshot();
  });

  it("shows an error alert if you have letters or symbols in the words", async () => {
    const stub = jest.fn();

    render(<AnagramRequestCard onSaveAnagram={stub} />);

    const firstInputField = screen.getByTestId("first-word-input");

    fireEvent.change(firstInputField, {
      target: { value: "racer22", name: "firstWord" },
    });

    expect(screen.getByTestId("error-alert")).toMatchSnapshot();
  });

  it("shows an error alert if you have symbols in the words", async () => {
    const stub = jest.fn();

    render(<AnagramRequestCard onSaveAnagram={stub} />);

    const firstInputField = screen.getByTestId("first-word-input");

    fireEvent.change(firstInputField, {
      target: { value: "!!abc", name: "firstWord" },
    });

    expect(screen.getByTestId("error-alert")).toMatchSnapshot();
  });

  it("shows an error alert if the anagram check request fails", async () => {
    const stub = jest.fn().mockRejectedValue("error with the request");

    render(<AnagramRequestCard onSaveAnagram={stub} />);

    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "racer", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "carer", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(screen.getByTestId("error-alert")).toMatchSnapshot();
  });
});
