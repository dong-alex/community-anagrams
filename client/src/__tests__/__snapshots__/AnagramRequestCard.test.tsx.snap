// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnagramRequestCard Tests renders the basic card 1`] = `
<div>
  <div
    class="card mt-5"
    data-test="card"
  >
    <div
      class="card-body"
      data-test="card-body"
    >
      <h4
        class="text-center mb-4"
      >
        Check Anagram
      </h4>
      <p
        class="text-center "
      >
        Check if two words are anagrams here.
         
        <strong>
          Do not include special characters and numbers.
        </strong>
      </p>
      <form>
        <div
          class="grey-text"
        >
          <div
            class="md-form"
          >
            
            <input
              aria-disabled="false"
              class="form-control"
              data-test="input"
              data-testid="first-word-input"
              name="firstWord"
              type="text"
              value=""
            />
            <label
              aria-labelledby=""
              class=""
              data-error=""
              data-success=""
              id=""
            >
              First Word
            </label>
          </div>
          <div
            class="md-form"
          >
            
            <input
              aria-disabled="false"
              class="form-control"
              data-test="input"
              data-testid="second-word-input"
              id="second-word-input"
              name="secondWord"
              type="text"
              value=""
            />
            <label
              aria-labelledby=""
              class=""
              data-error=""
              data-success=""
              for="second-word-input"
              id=""
            >
              Second Word
            </label>
          </div>
        </div>
        <div
          class="text-center"
        >
          <button
            class="btn-blue btn Ripple-parent disabled"
            data-test="button"
            data-testid="anagram-submission-button"
            disabled=""
            type="button"
          >
            Check
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
